File: java/Recursion_Backtracking/Letter_Combinations_of_Phone_Number.java
package java.Recursion_Backtracking;

import java.util.*;

public class Letter_Combinations_of_Phone_Number {
  String[] map = {
      "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
  };

  public List<String> letterCombinations(String digits) {
    List<String> result = new ArrayList<>();
    if (digits.length() == 0)
      return result;
    backtrack("", digits, 0, result);
    return result;
  }

  void backtrack(String current, String digits, int index, List<String> result) {
    if (index == digits.length()) {
      result.add(current);
      return;
    }
    String letters = map[digits.charAt(index) - '0'];
    for (char c : letters.toCharArray()) {
      backtrack(current + c, digits, index + 1, result);
    }
  }
}

File: C++/Recursion&Backtracking/Letter_Combinations_of_Phone_Number.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
  vector<string> mp = {
      "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

  void solve(string digits, int idx, string temp, vector<string> &ans)
  {
    if (idx == digits.size())
    {
      ans.push_back(temp);
      return;
    }
    string letters = mp[digits[idx] - '0'];
    for (char ch : letters)
      solve(digits, idx + 1, temp + ch, ans);
  }
  vector<string> letterCombinations(string digits)
  {
    vector<string> ans;
    if (digits.empty())
      return ans;
    solve(digits, 0, "", ans);
    return ans;
  }
};